test/e2e/autoscaling/horizontal_pod_autoscaling_behavior.go

```diff
@@ -212,6 +204,139 @@ var _ = SIGDescribe("[Feature:HPA] [Serial] [Slow] Horizontal pod autoscaling (n
 			replicas := rc.GetReplicas()
 			framework.ExpectEqual(replicas == initPods, true, "had %s replicas, still have %s replicas after time deadline", initPods, replicas)
 		})
+
 	})
 
+	ginkgo.Describe("with scale limited by number of Pods rate", func() {
+		ginkgo.It("should scale up no more than given number of Pods per minute", func() {
+			ginkgo.By("setting up resource consumer and HPA")
+			initPods := 1
+			initCPUUsageTotal := initPods * usageForSingleReplica
+			limitWindowLength := 1 * time.Minute
+			podsLimitPerMinute := 1
+
+			rc := e2eautoscaling.NewDynamicResourceConsumer(
+				hpaName, f.Namespace.Name, e2eautoscaling.KindDeployment, initPods,
+				initCPUUsageTotal, 0, 0, int64(podCPURequest), 200,
+				f.ClientSet, f.ScalesGetter, e2eautoscaling.Disable, e2eautoscaling.Idle,
+			)
+			defer rc.CleanUp()
+
+			hpa := e2eautoscaling.CreateCPUHorizontalPodAutoscalerWithBehavior(
+				rc, int32(targetCPUUtilizationPercent), 1, 10,
+				e2eautoscaling.HPABehaviorWithScaleLimitedByNumberOfPods(true, false, int32(podsLimitPerMinute), int32(limitWindowLength.Seconds())),
+			)
+			defer e2eautoscaling.DeleteHPAWithBehavior(rc, hpa.Name)
+
+			ginkgo.By("triggering scale up by increasing consumption")
+			rc.ConsumeCPU(4 * usageForSingleReplica)
+
+			ginkgo.By("assuring no more than 1 replica was added during 1 minute")
+			rc.EnsureDesiredReplicasInRange(1, 2, limitWindowLength, hpaName)
+
+			ginkgo.By("assuring no more than 2 replicas were added during 2 minutes")
+			rc.EnsureDesiredReplicasInRange(1, 3, limitWindowLength, hpaName)
+
+			ginkgo.By("assuring no more than 3 replicas were added during 3 minutes")
+			rc.EnsureDesiredReplicasInRange(1, 4, limitWindowLength, hpaName)
```

Comment:
If HPA doesn't scale at all due to a bug then this test still passes (1 is in range [1, 4]). I would prefer to verify that we really went from 1 replica to 2 replicas after a minute, similar to what we do in stabilization tests. WDYT?

(also applies do downscale test and percentage tests)

Reply:
To save test run time we can probably get away with just a single scale up/down verification as opposed to 2 or 3 in a scenario.

Reply:
As discussed offline: I've refactored this totally. We need to have two steps for each, as we can't have any reasonable contraints for the first scale event - it can happen within window limit, but can also happen later, due to various delays. In order to model this, IMO it's the best to have two steps and measure the time it took to scale up/down before each of them. Then we can compare if first one happened before deadline and second happened after window limit length and before deadline.